tests/semantic_analyser.cpp

```diff
@@ -4035,4 +4035,78 @@ kfunc:foo { $p = path((uint8 *)0) }
   test("BEGIN { $b = buf(0, 10) }");
 }
 
+TEST(semantic_analyser, variable_declarations)
+{
+  test("BEGIN { let $a; $a = 1; }");
+  test("BEGIN { let $a: int16; $a = 1; }");
+  test("BEGIN { let $a = 1; }");
+  test("BEGIN { let $a: int16 = 1; }");
+  test("BEGIN { let $a: str_t[5] = \"hiya\"; }");
+  // If the type is specified it's strict in that future assignments
+  // need to fit into that type
+  test(R"(BEGIN { let $a: str_t[5] = "hiya"; $a = "bye"; })");
+  test(R"(BEGIN { let $a = "hiya"; $a = "longerstr"; })");
+  test("BEGIN { let $a: int16 = 1; $a = (int8)2; }");
+  // Test more types
+  test("BEGIN { let $a: struct x; }");
+  test("BEGIN { let $a: struct x *; }");
```

Comment:
This caught my attention (only now, sorry!) as it can introduce a lot of tricky things.

When I tried it, though, I'm unable to get it past semantic analyser:
```
# bpftrace -e 'BEGIN { let $f: struct task_struct; }'
stdin:1:17-38: ERROR: syntax error, unexpected end of file, expecting identifier
BEGIN { let $f: struct task_struct; }
                ~~~~~~~~~~~~~~~~~~~~~

# bpftrace -e 'BEGIN { let $f: struct task_struct *; }'
stdin:1:17-40: ERROR: syntax error, unexpected end of file, expecting identifier
BEGIN { let $f: struct task_struct *; }
                ~~~~~~~~~~~~~~~~~~~~~~~
```
Is that the expected behaviour? And if so, how come that the tests pass?

Reply:
@viktormalik Are you on an older version of this PR? This was the behavior before I updated the lexer.

Reply:
Indeed I was and when I tried to update, I accidentally broke your PR :sob: I'm so sorry about that, you'll need to force-push the latest version.

Reply:
Anyways, with the latest version of this PR, I revealed a couple of bugs:

1. Variable of struct type:
```
# bpftrace -e 'BEGIN { let $f: struct task_struct; $f = *curtask; print($f.pid); }'
Segmentation fault (core dumped)
```
While without the declaration, it correctly reports the size of `task_struct` exceeding the stack:
```
# bpftrace -e 'BEGIN { $f = *curtask; print($f.pid); }'
stdin:1:9-22: ERROR: Value is too big (13696 bytes) for the stack. Try reducing its size, storing it in a map, or creating it in argument position to a helper call.

Examples:
    `$s = str(..);` => `$s = str(.., 32);`
    `$s = str(..);` => `@s = str(..);`
    `$s = str(..); print($s);` => `print(str(..));`

BEGIN { $f = *curtask; print($f.pid); }
        ~~~~~~~~~~~~~

```

2. Variable of struct pointer type:
```
bpftrace -e 'BEGIN { let $f: struct task_struct *; $f = curtask; print($f->pid); }'
stdin:1:40-52: ERROR: Type mismatch for $f: trying to assign value of type 'struct task_struct *' when variable already has a type 'struct task_struct *'
BEGIN { let $f: struct task_struct *; $f = curtask; print($f->pid); }
                                       ~~~~~~~~~~~~
stdin:1:54-62: ERROR: Variable used before it was assigned: $f
BEGIN { let $f: struct task_struct *; $f = curtask; print($f->pid); }
                                                     ~~~~~~~~
```

Reply:
For the record, I'm perfectly ok with disabling declarations of struct-typed variables now and fixing them later.

Reply:
Good catches! I found fixes for both.

Reply:
Yep, looks good, thanks!